bip-generic-signmessage.mediawiki

```diff
@@ -0,0 +1,134 @@
+# Derive the private key privkey for the scriptPubKey, or FAIL
+# Define the message pre-image as the sequence "Bitcoin Message:" concatenated with the message, encoded in UTF-8 using Normalization Form Compatibility Decomposition (NFKD)
+# Let sighash = sha256(sha256(pre-image))
+# Generate a signature sig with privkey=privkey, sighash=sighash
+
+Repeat the above operation for each scriptPubKey, retaining the scriptPubKeys set. As noted, if the same scriptPubKey appears more than once, the sign operation must fail.
+
+=== Verifying ===
+
+The "Verify" action takes as input a standard flags value, a scriptPubKey, a message, a script sig, and a witness program.
+It emits one of INCONCLUSIVE, VALID, INVALID, or ERROR.
+
+# Return ERROR if scriptPubKey already exists in scriptPubKeys set, otherwise insert it
+# If one or more of the standard flags are unknown, return INCONCLUSIVE
+# Define the message pre-image as the sequence "Bitcoin Message:" concatenated with the message, encoded in UTF-8 using Normalization Form Compatibility Decomposition (NFKD).
+# Let sighash = sha256(sha256(pre-image))
+# Verify Script with flags=standard flags, scriptSig=script sig, scriptPubKey=scriptPubKey, witness=witness program, and sighash=sighash
+# Return VALID if verify succeeds, otherwise return INVALID
+
+Repeat the above operation for each scriptPubKey, retaining the scriptPubKeys set. As noted, if the same scriptPubKey appears more than once, the verify operation must fail with an ERROR.
+
+* If a verification call returns ERROR or INVALID, return ERROR or INVALID immediately, ignoring as yet unverified entries.
+* After all verifications complete, return INCONCLUSIVE if any verification call returned INCONCLUSIVE.
+* Return VALID if and only if every verification returned VALID.
+
+== Future Extensions ==
+
+=== Proof of Funds ===
+
+The specification can be extended to handle proof of funds in the following manner:
+
+* Let the message be prefixed with "POF:", followed by a newline-terminated string<ref><strong>Why not just the UTXO data?</strong> We want the verifier to be able to challenge the prover with a custom message to sign, or anyone can reuse the POF proof for a set of UTXO:s once they have seen it, and the funds have not yet been spent</ref>, followed by [entries] series of hex-encoded transaction ID:vout pairs, separated by a single space (" ") character
+* Fail if the number of txid:vout pairs is not exactly equal to [entries]
+* Retain the message as is for all sighash operations (i.e. all sign and verify operations sign and verify for the entire list of UTXO:s)<ref><strong>Why use same sighash?</strong> The prover is proving that they have a set of UTXO:s at their disposal. Taking a sub-set of the proofs and turning into a new proof should not be valid.</ref>
```

Comment:
Suggested improvement:
Line 108: Retain the message as is for all sighash operations (i.e. all sign and verify operations **should** sign and **verify the entire list** of UTXO:s ...........Taking a sub-set of the proofs and turning **them** into a new proof should not be valid.

or 
Line 108: Retain the message as is for all sighash operations (i.e. all sign and verify operations **should validate the entire list** of UTXO:s ...........Taking a sub-set of the proofs and turning **them** into a new proof should not be valid.